stable_baselines3/common/base_class.py

```diff
@@ -435,6 +436,15 @@ def _setup_learn(
         :param progress_bar: Display a progress bar using tqdm and rich.
         :return: Total timesteps and callback(s)
         """
+
+        if eval_env is not None or eval_freq != -1:
+            warnings.warn(
+                "Parameters `eval_env` and `eval_freq` are deprecated and will be removed in the future."
+                " Please use `EvalCallback` or a custom Callback instead.",
+                DeprecationWarning,
+                stacklevel=4,
```

Comment:
good catch. There are two reasons why I choose `stacklevel=4`
1. To refer to the warning where the deprecated feature was used (the call of `model.learn()`). Therefore, the warning, in the end, looks like this:
```
DeprecationWarning: Parameters `eval_env` and `eval_freq` are deprecated and will be removed in the future. Please use `EvalCallback` or a custom Callback instead.
  model.learn(total_timesteps=10000, eval_freq=100)
```

and not like this (if `stacklevel=1`, which is the default value):
```
Parameters `eval_env` and `eval_freq` are deprecated and will be removed in the future. Please use `EvalCallback` or a custom Callback instead.
  warnings.warn(
```
2. By default, warnings of the category `DeprecationWarning` are ignored, unless triggered by code in `__main__` (See [python docs](https://docs.python.org/3/library/warnings.html#warning-categories)). By setting `stacklevel=4`, we are referring to the initial call (`model.learn()`), which is mainly triggered by code in `__main__`. Therefore, we can show the warning without changing the behavior of the default warning filter.

Alternatively to setting `stacklevel=4`, we could change the default warning filter like (see [python docs](https://docs.python.org/3/library/warnings.html#overriding-the-default-filter)):
```
import sys

if not sys.warnoptions:
    import warnings
    warnings.simplefilter("ignore")
    warnings.warn('do the warning here`)
    warnings.simplefilter("default")
```

- An argument for using the `stacklevel` argument is: We don't need to change any default behaviour and ease of implementation.
- An argument against `stacklevel` argument is: The stacklevel is static. If the call does not refer to `model.learn()` "in level 4", or `model.learn()` is not called in `__main__` the warning is "swallowed"

In the end, we need to go for one of both if we want to show the warning when the parameter is used when using sb3 (In order to inform developers about the deprecated parameter). Otherwise, the warning is simply swallowed most of the time and only shown in rare cases e.g. when running the tests. 

Please let me know if I got something wrong or if I should change it from using the `stacklevel` parameter to changing the warning filter.

Reply:
Thank you for the detailed explanation. I agree with your current use of stack level.
Can you add a comment line so that we remember why `stack_level = 4`?